### PR TITLE
Retry curl when it failed during download

### LINKED ISSUE #34769: Curl e2e flake
Sometimes e2e tests fails during cluster installation because curl fails with:
```
curl: (18) transfer closed with 4980557 bytes remaining to read
+++ exit code: 18
+++ error: 1
```
For example see here: https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/pr-logs/pull/32781/kubernetes-pull-test-unit-integration/48792/ Downloading etcd failed.

I've run curl to get etcd from github 4400 times. It failed 5 times. So it doesn't happen often.

### PR DESCRIPTION
**Pull request by lukaszo**
Fixes #34769

For now I'm adding retry only to etcd installation as I saw it recently to fail in one of my PRs. We may change it also in other places if it also happens.

---

This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/34770)

### CODE REVIEW

**Comment by spxtr** [comment]
I think @ixdy was looking at this stuff recently. Were we going to put this in the test image?

**Comment by ixdy** [comment]
Maybe eventually, but that depends on us using the new images (which we aren't yet).

File: hack/lib/util.sh [hunk 1]
```diff
@@ -478,4 +478,25 @@ kube::util::has_changes_against_upstream_branch() {
...
 }
 
+kube::util::download_file() {
+  local -r url=$1
+  local -r destination_file=$2
+
+  rm  ${destination_file} >/dev/null || true
+
+  n=0
+  for i in {1..5}
```
Comment (spxtr) [inline hack/lib/util.sh]: `for i in $(seq 5); do`

File: hack/lib/util.sh [hunk 2]
```diff
@@ -478,4 +478,25 @@ kube::util::has_changes_against_upstream_branch() {
   return 1
 }
 
+kube::util::download_file() {
+  local -r url=$1
+  local -r destination_file=$2
+
+  rm  ${destination_file} >/dev/null || true
```
Comment (spxtr) [inline hack/lib/util.sh]: `2&> /dev/null`, since `rm` puts the no such file message to stderr.

File: hack/lib/util.sh [hunk 3]
```diff
@@ -478,4 +478,25 @@ kube::util::has_changes_against_upstream_branch() {
...
+
+  n=0
+  for i in {1..5}
+  do
+    if ! curl -fsSL --retry 3 --keepalive-time 2 ${url} -o ${destination_file}; then
+      n=$((n+1))
+      echo "Downloading '${url}' failed. Retrying."
+      echo "$((5-i)) retries left."
```
Comment (spxtr) [inline hack/lib/util.sh]: I would put these two messages on the same line.

**Review by spxtr — COMMENTED** [review]
Thanks for the change. I have some cosmetic comments, but it looks good.

File: hack/lib/util.sh (hunk 2 above)
Comment (lukaszo) [inline hack/lib/util.sh]: Done

File: hack/lib/util.sh (hunk 1 above)
Comment (lukaszo) [inline hack/lib/util.sh]: done

File: hack/lib/util.sh (hunk 3 above)
Comment (lukaszo) [inline hack/lib/util.sh]: done

**Comment by lukaszo** [comment]
@k8s-bot gci gke e2e test this
@k8s-bot cvm gke e2e test this

Looks like problems with starting cluster

**Comment by lukaszo** [comment]
@spxtr I updated the code after your comments. PTAL

File: hack/lib/util.sh [hunk 4]
```diff
@@ -478,4 +478,22 @@ kube::util::has_changes_against_upstream_branch() {
...
+  rm  ${destination_file} 2&> /dev/null || true
+
+  for i in $(seq 5)
+  do
+    if ! curl -fsSL --retry 3 --keepalive-time 2 ${url} -o ${destination_file}; then
+      echo "Downloading ${url} failed. $((5-i)) retries left."
+      sleep 1
+    else
+      echo "Downloading ${url} succeed"
+      break
```
Comment (spxtr) [inline hack/lib/util.sh]: Maybe make this return 0 and if we exit the loop normally then return 1?

**Comment by spxtr** [comment]
LG other than comment.

**Comment by lukaszo** [comment]
@k8s-bot gci gke e2e test this

again problems with starting cluster

**Comment by lukaszo** [comment]
@k8s-bot gci gke e2e test this

**Comment by lukaszo** [comment]
@k8s-bot gci gke e2e test this #35123

**Comment by spxtr** [comment]
@k8s-bot gci gke e2e test this #35123

**Pull request merged**